### ISSUE TITLE
A Visual Tour of Backpropagation

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11755]
A Visual Tour of Backpropagation<br>
<br>
https://ift.tt/5GqATmH<br>
<br>
Jinay Jain<br>
<br>

<div><div><div><div>
<div>
<p>Before reading, you can try out my <a href="https://jinay.dev/backprop-vis" target="_blank">interactive visualization
tool</a> for a primer on what I’ll discuss
below.</p>
<p>At their core, neural networks are functions. They take some input, perform a
series of computations, and produce an output. Though most networks operate
in the realm of vectors and matrices, it can be a useful exercise to see them
without the extra barrier of linear algebra. For this purposes of this
explanation, we will only cover single variable functions, but the principles
we will see can be extended into any number of dimensions.</p>
<h1>The Forward Pass</h1>
<p>Before we reach the backwards propagation of gradients, we will observe the
forward propagation of values. The forward pass provides a reasonable basis
for understanding how functions can be represented as the result of other
functions combined. Finally, we can represent these functions as a
computation graph—an idea which will be useful when we arrive at the
backward pass of the function.</p>
<h2>Function Composition</h2>
<p>We begin with the most basic function, mapping the entire range of inputs to a single number.</p>
<p><span><span><span></span></span></span></p>
<p>Simple, right? Let’s try to incorporate <span><span></span></span> into the mix by creating the identity function—any <span><span></span></span> in will be the same <span><span></span></span> out.</p>
<p><span><span><span></span></span></span></p>
<p>The power comes when we can compose the functions using different operations,
building a new function from the existing ones. By composing functions
together, we can evaluate the individual functions before we evaluate the
parent.</p>
<p><span><span><span></span></span></span></p>
<p>Likewise, we can feed the output of one function as the input of another,
increasing the complexity of the overall function while retaining its
individual parts.</p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p>Intuitively, a human would first calculate the value of <span><span></span></span> before
squaring the result. This staged ordering creates a hierarchy of functions
that our algorithms can harness.</p>
<p>We see this type of hierarchy building when we build models in any machine learning library. What TensorFlow calls a “model” is just a series of computations composed into a single function.</p>
<pre><code>model = tf.keras.Sequential()

model.add(Dense(<span>512</span>))
model.add(Dense(<span>512</span>))
...
model.add(Dense(<span>32</span>))
model.add(Dense(<span>10</span>))

y = model.predict(X)
</code></pre>
<p>Every time we call <code>model.add()</code>, we are adding to the hierarchy that defines
our neural network. In machine learning, composed functions <span><span></span></span> defined
earlier are less like <span><span></span></span> and more like</p>
<p><span><span><span></span></span></span></p>
<p>Each <span><span></span></span> is a layer in the network which produces an output by using the previous layer as input.</p>
<h2>Building a Computation Graph</h2>
<p>We can visualize a composed function’s structure as a tree, each layer
representing a different stage of the operation. In fact, most programming
languages use a <a href="https://en.wikipedia.org/wiki/Parse_tree" target="_blank">parse tree</a> to
store and evaluate expressions in code. This representation gives us the
ability to view a function in terms of its component parts, recursing
down the levels until we find the constants and variables at the
very bottom.</p>
<p>For example, take the function <span><span></span></span> given below.</p>
<p><span><span><span></span></span></span></p>
<p>In machine learning, this is a function called the <em>sigmoid</em>, often written
as <span><span></span></span>. Let’s see how we could break the sigmoid into a computation
tree.</p>
<p><img alt="sigmoid" src="https://blog.jinay.dev/images/sigmoid1.svg"></p>
<p>Notice how all four leaf nodes are either a constant or a variable. These
are the most elementary parts of any composite function. We can even give the
intermediate functions letter names to highlight how each node is built from its children.</p>
<p><img alt="sigmoid with labels" src="https://blog.jinay.dev/images/sigmoid2.svg"></p>
<p>The equations for our new tree would be</p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<h1>The Backward Pass</h1>
<p>Up until this point, viewing functions as hierarchies seems to add
unnecessary abstractions on a relatively simple topic. However, now that we have
observed how to compute the “forward pass” of a computation graph, we
can compute the reverse using an algorithm called
<a href="https://en.wikipedia.org/wiki/Backpropagation" target="_blank">backpropagation</a>.</p>
<h2>Gradient Descent</h2>
<p>In machine learning, the end goal is to <strong>minimize error on a loss
function</strong>, and modern software achieves this goal through an algorithm called
<em>gradient descent</em>. Though I will not attempt to explain the entirety of
gradient descent here, a basic understanding of how it works is essential for
understanding backpropagation.</p>
<p>Loss functions measure how much the outputs of a model, the neural network,
deviate from the labels in a dataset. Tuning the parameters of the model will
either increase or decrease that loss, and the goal is to find the set of
parameters that will give us the minimum loss. Through gradient descent, we
try to estimate which direction we should tune our model in order to achieve
the optimal settings.</p>
<p>Imagine we have a loss function <span><span></span></span> that measures the performance of a model with only one parameter, <span><span></span></span>. The graph of <span><span></span></span> could look something like this graph:</p>
<p><img src="https://blog.jinay.dev/images/graph1.png"></p>
<p>From this zoomed out perspective, the valley in the curve (where loss is the
lowest) is obvious. However, when we train a model, this view is
significantly smaller, giving us only information about how the loss curve looks
near the current value of <span><span></span></span>.</p>
<p><img src="https://blog.jinay.dev/images/graph2.png"></p>
<p>Intuitively, we should follow the direction that has the steepest slope downwards to find the minimum. In mathematical terms, we want to look at the gradient of <span><span></span></span> and take a small step in the direction down that gradient.</p>
<p><img src="https://blog.jinay.dev/images/graph3.png"></p>
<p><em>Backpropagation</em> is the tool that helps a model find that gradient estimate so that we know which direction to move in.</p>
<h2>Backpropagation</h2>
<p>The gradient is a collection of slope calculations called the partial
derivatives. Both partial derivatives and gradients answer a question that is
fundamental to our purpose: how does a small change in a variable <span><span></span></span>
respectively change the output function <span><span></span></span>. In machine learning, we want
to observe how changing <span><span></span></span> will change the loss.</p>
<p>Before we begin, let us revisit some of the basic rules of calculus that are
crucial to understanding backpropagation. When computing partial derivatves,
we consider other variables of the functions constants, so <span><span></span></span> represents any constant or variable other than <span><span></span></span>.</p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p>There are many more rules, but these basics encompass a large portion of what
functions you might see in machine learning applications. However,
using only these rules, our function vocabulary is significantly
limited. We can only take partial derivatives for simple functions like <span><span></span></span>
or <span><span></span></span> if we are clever. The essential property for achieving infinitely
complex functions is a rule most calculus classes group with the rest.</p>
<p><span><span><span></span></span></span></p>
<p>The chain rule enables us to unravel the composed functions we discussed
earlier—giving us the ability to compute arbitrarily complex partial
derivatives and gradients. Using the computation graph we constructed
earlier, we can move backwards from the final result to find individual
derivatives for variables. This backwards computation of the derivative using
the chain rule is what gives backpropagation its name. We use the
<span><span></span></span> and propagate that partial derivative
backwards into the children of <span><span></span></span> .</p>
<p>As a simple example, consider the following function and its corresponding computation graph.</p>
<p><span><span><span></span></span></span></p>
<p><img alt="simple example" src="https://blog.jinay.dev/images/backprop1.svg"></p>
<p>As earlier, we assign the intermediate computations letter names to use in
calculation. Let’s find <span><span></span></span>, the slope of <span><span></span></span>,
when <span><span></span></span>. First, let’s compute the forward pass of the function, labelling
the intermediate values of each node until we reach <span><span></span></span>.</p>
<p>Starting from the top, let us find the partial derivatives of the
tree on the path down to <span><span></span></span>. As we traverse down the tree, it only makes sense to take the partial derivatives with respect to the branches that <span><span></span></span> is a part of.</p>
<p><span><span><span></span></span></span></p>
<p>Since <span><span></span></span> is a direct child of g, the chain rule is not necessary here. Knowing the values of <span><span></span></span> and <span><span></span></span> from the forward pass, we can find a numerical value for the partial derivative.</p>
<p><span><span><span></span></span></span></p>
<p>The next node in the path is <span><span></span></span>, a direct child of the previous node <span><span></span></span>. To
find the “local” partial derivative, <span><span></span></span> we use
the addition rule. <span><span></span></span> is 0 since a small change in <span><span></span></span> makes no change in <span><span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p>To find the desired value of <span><span></span></span>, we can finally employ the chain rule.</p>
<p><span><span><span></span></span></span></p>
<p>Finally, we arrive at node <span><span></span></span>, the node containing <span><span></span></span>. The process for finding this derivative is identical to the previous ones.</p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p><span><span><span></span></span></span></p>
<p>We’ve arrived at the end of the path to <span><span></span></span>, giving us the final value of <span><span></span></span>.</p>
<p>I encourage you to revisit my <a href="https://jinay.dev/backprop-vis/" target="_blank">visualization
tool</a> and play around with various
functions, tracing the path of the gradients down to your original variables.
You might even want to try <code>(4*x+1)^2</code> and see how the gradients change as
<span><span></span></span> changes.</p>
<p>This simple algorithm for calculating partial derivatives on a computation
graph is very similar to the way neural networks are trained in libraries
like Tensorflow. A firm understanding on how the libraries work allows you to
expand your capabilities beyond the prepackaged layers they provide. More
importantly, it allows you to gain insight on the algorithms used every day to train
computers on any amount of data.</p>
<p>The task remains constant—minimize error on a loss function.</p>
<h3>Additional Resources</h3>
<p><em>This post was inspired by the insightful explanations of backpropagation linked below.</em></p>
<p><a href="https://cs231n.github.io/optimization-2/" target="_blank">Stanford CS231n - Backpropagation, Intuitions</a></p>
<p><a href="https://www.youtube.com/watch?v=Ilg3gGewQ5U" target="_blank">3Blue1Brown - What is backpropagation really doing?</a></p>
<p><a href="https://www.youtube.com/watch?v=GlcnxUlrtek" target="_blank">Welch Labs - Neural Networks Demystified - Backpropagation</a></p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Jinay Jain - Blog <br>
<br>
August 22, 2024 at 03:57PM
